SpockOpaleye/src/Email.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes       #-}
+
+module Email where
+
+import           Network.Mail.Mime
+import           Network.Mail.SMTP  hiding (simpleMail)
+
+import           ApiBase
+import           Conf               (apikey)
+import           Control.Concurrent
+import           Control.Lens
+import           Data.ByteString    hiding (putStrLn)
+import           Data.Monoid
+import           Data.String
+import           Data.String.Here
+import qualified Data.Text          as T
+import qualified Data.Text.Lazy     as LT
+import           DataTypes
+
+sendgridMail :: Mail -> IO ()
+sendgridMail mail = do
+  threadId <- forkIO $ sendMailWithLogin' "smtp.sendgrid.net" 587 "apikey" apikey mail
+  putStrLn $ show $ T.concat ["Sending main from thread ", fromString $ show threadId]
+
+setCid :: T.Text -> T.Text -> Mail -> Mail
+setCid filename cid mail@Mail {mailParts = alternatives} = mail { mailParts = (setCidInAlternative filename cid) <$> alternatives}
+  where
+  setCidInAlternative :: T.Text -> T.Text -> Alternatives -> Alternatives
+  setCidInAlternative fp cid parts = (setCidForPart filename cid) <$> parts
+  setCidForPart :: T.Text -> T.Text -> Part -> Part
+  setCidForPart filename cid p@Part{partFilename = Just fn, partHeaders = ph} =
+    if fn == filename then p{partHeaders = (makeContentIdHeader cid):ph} else p
+  setCidForPart filename cid  p = p
+  makeContentIdHeader :: T.Text -> (ByteString, T.Text)
+  makeContentIdHeader cid = ("Content-ID", T.concat ["<", cid,  ">"])
+
+addLogo :: Mail -> IO Mail
+addLogo mail = do
+  mail_with_attachment <- addAttachment "image/png" "apple.png" mail
```

Comment:
That is right. I didn't want to reimplement the attachment creation using add part because I was not sure what hard that would be to do reliably. So I used the addAttachment function and then walk through the resulting mail and look for the filename we just added, and add the content id header to that `Part`.

Now that we have seen the code of the addAttachment function, it seems quite trivial enough to just use addPart and make the attachment ourselves.